akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/HttpCookie.scala

```diff
@@ -55,16 +57,75 @@ object HttpCookiePair {
  * for a full definition of the http cookie header fields, see
  * http://tools.ietf.org/html/rfc6265
  */
-final case class HttpCookie(
-  name:      String,
-  value:     String,
-  expires:   Option[DateTime] = None,
-  maxAge:    Option[Long]     = None,
-  domain:    Option[String]   = None,
-  path:      Option[String]   = None,
-  secure:    Boolean          = false,
-  httpOnly:  Boolean          = false,
-  extension: Option[String]   = None) extends jm.headers.HttpCookie with ToStringRenderable {
+final class HttpCookie private[http] (
+  name:          String,
+  value:         String,
+  val expires:   Option[DateTime],
+  val maxAge:    Option[Long],
+  val domain:    Option[String],
+  val path:      Option[String],
+  secure:        Boolean,
+  httpOnly:      Boolean,
+  val extension: Option[String],
+  val sameSite:  Option[SameSite]) extends jm.headers.HttpCookie with ToStringRenderable with Product with Serializable with Equals {
+
+  @deprecated("for binary compatibility", since = "10.2.0")
+  def this(
```

Comment:
we can make this private[headers]

Reply:
Making this constructor `private[headers]` (and also the deprecated `apply` method) will both have an impact on binary compatibility. I can do that and configure the filters, but want to check with your first before doing that.

Reply:
Ah, you're right, because of of the default parameters and the change from `case class` to `class`... In that case I agree it makes sense to keep the constructor with default parameters separate like this, but then perhaps it should be public and not deprecated?

Reply:
I wonder if we can instead target source compatibility here instead of binary compatibility. If so, we can have a single constructor with default params and accepting `Option[SameSite]` (defaulting to `None`).

The alternative is to either keep this constructor without the deprecation (which does not add value to the API but creates some confusion if I want to create a cookie with `SameSite`) or keep it deprecated and go through the cycle of removing it later and adding the default value in the main constructor.

WDYT?

Reply:
By the way, going for source compatibility would require the following new filters:

```
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$9")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$8")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$7")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$6")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$5")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$4")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.<init>$default$3")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.model.headers.HttpCookie.this")
```

Just to make the impact clearer.

Reply:
I don't think we can afford to skip binary compatibility here - any library that uses the current HttpCookie constructor is using those methods for the default values, so if we'd break binary compatibility all such libraries would start throwing exceptions.

> The alternative is to either keep this constructor without the deprecation (which does not add value to the API but creates some confusion if I want to create a cookie with `SameSite`)

I agree it is a bit confusing that you'd have to use `withSameSite` for sameSite instead of using this constructor, but for all other uses it adds value by providing a way to create a HttpCookie without providing values for all fields.

> keep it deprecated and go through the cycle of removing it later and adding the default value in the main constructor

Unforunately you can't have 2 constructors that both have default values

Reply:
> In that case I agree it makes sense to keep the constructor with default parameters separate like this, but then perhaps it should be public and not deprecated?

@raboof I went this route. This has a little conflict with what @jrudolph suggested before (make the constructor `private[headers]`).

Reply:
Do we still want to keep this public and un-deprecated? I think we should at least deprecate it.

```suggestion
  @deprecated("Use apply method, HttpCookie(...), instead of new HttpCookie(...)", since = "10.2.0")
  def this(
```